source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.h

```diff
@@ -84,36 +84,55 @@ using GrpcClientImplPtr = std::unique_ptr<GrpcClientImpl>;
 // The client cache for RawAsyncClient for Google grpc so channel is not created for each request.
 // TODO(fpliu233): The cache will cause resource leak that a new channel is created every time a new
 // config is pushed. Improve gRPC channel cache with better solution.
-class AsyncClientCache : public Singleton::Instance {
+class AsyncClientCache : public std::enable_shared_from_this<AsyncClientCache> {
 public:
   AsyncClientCache(Grpc::AsyncClientManager& async_client_manager, Stats::Scope& scope,
                    ThreadLocal::SlotAllocator& tls)
-      : async_client_manager_(async_client_manager), scope_(scope), tls_slot_(tls.allocateSlot()) {
-    tls_slot_->set([](Event::Dispatcher&) { return std::make_shared<ThreadLocalCache>(); });
-  }
-
-  const Grpc::RawAsyncClientSharedPtr getOrCreateAsyncClient(
-      const envoy::extensions::filters::http::ext_authz::v3::ExtAuthz& proto_config);
+      : async_client_manager_(async_client_manager), scope_(scope),
+        tls_slot_(ThreadLocal::TypedSlot<ThreadLocalCache>::makeUnique(tls)) {}
```

Comment:
You don't need the extra pointer indirection in this case. You can make tls_slot_ be a `ThreadLocal::TypedSlot` (no Ptr) and initialize it as:
```
  tls_slot_(tls)
```

Then you can change `tls_slot_->get().foo` to `tls_slot_->foo` and otherwise `tls_slot_->` to `tls_slot_.` throughout.

Reply:
Done. Thanks for the guidance!